pilot/pkg/xds/workloadentry.go

```diff
@@ -0,0 +1,219 @@
+	ConnectedAtAnnotation = "istio.io/connectedAt"
+	// DisconnectedAtAnnotation on a WorkloadEntry stores the time in nanoseconds when the associated workload disconnected from a Pilot instance.
+	DisconnectedAtAnnotation = "istio.io/disconnectedAt"
+)
+
+func (sg *InternalGen) RegisterWorkload(proxy *model.Proxy, con *Connection) {
+	// check if the WE already exists, update the status
+	entryName := autoregisteredWorkloadEntryName(proxy)
+	if entryName == "" {
+		return
+	}
+
+	// Try to patch, if it fails then try to create
+	_, err := sg.Store.Patch(gvk.WorkloadEntry, entryName, proxy.Metadata.Namespace, func(cfg config.Config) config.Config {
+		setConnectMeta(&cfg, sg.Server.instanceID, con)
+		return cfg
+	})
+	if err == nil {
+		return
+	}
+	// TODO better errors from Get/Patch. NotFound is fine but others should warn.
+
+	// No WorkloadEntry, create one using fields from the associated WorkloadGroup
+	groupCfg := sg.Store.Get(gvk.WorkloadGroup, proxy.Metadata.AutoRegisterGroup, proxy.Metadata.Namespace)
+	if groupCfg == nil {
+		adsLog.Warnf("auto registration of %v failed: cannot find WorkloadGroup %s/%s", proxy.ID, proxy.Metadata.Namespace, proxy.Metadata.AutoRegisterGroup)
+		return
+	}
+	entry := workloadEntryFromGroup(entryName, proxy, groupCfg)
+	setConnectMeta(entry, sg.Server.instanceID, con)
+	_, err = sg.Store.Create(*entry)
+	if err != nil {
+		adsLog.Errorf("auto registration of %v failed: error creating WorkloadEntry: %v", proxy.ID, err)
+	}
+}
+
+func (sg *InternalGen) QueueUnregisterWorkload(proxy *model.Proxy) {
+	// check if the WE already exists, update the status
+	entryName := autoregisteredWorkloadEntryName(proxy)
+	if entryName == "" {
+		return
+	}
+	// unset controller, set disconnect time
+	_, err := sg.Store.Patch(gvk.WorkloadEntry, entryName, proxy.Metadata.Namespace, func(cfg config.Config) config.Config {
+		delete(cfg.Annotations, WorkloadControllerAnnotation)
+		cfg.Annotations[DisconnectedAtAnnotation] = strconv.FormatInt(time.Now().UnixNano(), 10)
+		return cfg
+	})
+	if err != nil {
+		adsLog.Warnf("disconnect: failed patching WorkloadEntry %s/%s: %v", proxy.Metadata.Namespace, entryName, err)
+		return
+	}
+
+	gp := time.After(features.WorkloadEntryCleanupGracePeriod)
+	go func() {
+		<-gp
+		sg.delayedCleanup <- config.Meta{Name: entryName, Namespace: proxy.Metadata.Namespace}
+	}()
+
+}
+
+// workloadEntryCleanup processes the delete queue as well as performs periodic cleanup of WorkloadEntries to catch
+// edge cases where the last connected pilot cannot perform the cleanup.
+func (sg *InternalGen) workloadEntryCleanup(stopCh <-chan struct{}) {
+	ticker := time.NewTicker(10 * features.WorkloadEntryCleanupGracePeriod)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			for _, ns := range sg.Store.Namespaces() {
+				wles, err := sg.Store.List(gvk.WorkloadEntry, ns.Name)
+				if err != nil {
+					continue
+				}
+				for _, wle := range wles {
+					go sg.cleanupEntry(wle)
```

Comment:
Maybe we should have a queue, otherwise we will have leaking goroutines if it gets backed up and ordering will not be "fair" I think (unless the ratelimitter is implicitly handling ordering)